src/USER-PHONON/dynamical_matrix.cpp

```diff
@@ -0,0 +1,491 @@
+{
+    if (narg < 0) error->all(FLERR,"Illegal dynamical_matrix command");
+    int iarg = 0;
+    const char* filename = "dynmat.dyn";
+    while (iarg < narg) {
+        if (strcmp(arg[iarg],"binary") == 0) {
+            if (iarg + 2 > narg) error->all(FLERR, "Illegal dynamical_matrix command");
+            if (strcmp(arg[iarg+1],"gzip") == 0) {
+                compressed = 1;
+            }
+            else if (strcmp(arg[iarg+1],"yes") == 0) {
+                binaryflag = 1;
+            }
+            iarg += 2;
+        }
+        else if (strcmp(arg[iarg],"file") == 0) {
+            if (iarg+2 > narg) error->all(FLERR, "Illegal dynamical_matrix command");
+            filename = arg[iarg + 1];
+            file_flag = 1;
+            iarg += 2;
+        } else error->all(FLERR,"Illegal dynamical_matrix command");
+    }
+    if (file_flag == 1) {
+        openfile(filename);
+    }
+}
+
+/* ----------------------------------------------------------------------
+   generic opening of a file
+   ASCII or binary or gzipped
+   some derived classes override this function
+------------------------------------------------------------------------- */
+
+void DynamicalMatrix::openfile(const char* filename)
+{
+    // if file already opened, return
+    if (me!=0) return;
+    if (file_opened) return;
+
+    if (compressed) {
+#ifdef LAMMPS_GZIP
+        char gzip[128];
+    sprintf(gzip,"gzip -6 > %s",filename);
+#ifdef _WIN32
+    fp = _popen(gzip,"wb");
+#else
+    fp = popen(gzip,"w");
+#endif
+#else
+        error->one(FLERR,"Cannot open gzipped file");
+#endif
+    } else if (binaryflag) {
+        fp = fopen(filename,"wb");
+    } else {
+        fp = fopen(filename,"w");
+    }
+
+    if (fp == NULL) error->one(FLERR,"Cannot open dump file");
+
+    file_opened = 1;
+}
+
+/* ----------------------------------------------------------------------
+   create dynamical matrix
+------------------------------------------------------------------------- */
+
+void DynamicalMatrix::calculateMatrix()
+{
+    int local_idx; // local index
+    int local_jdx; // second local index
+    int nlocal = atom->nlocal;
+    bigint natoms = atom->natoms;
+    int *mask = atom->mask;
+    int *type = atom->type;
+    double imass; // dynamical matrix element
+    double *m = atom->mass;
+    double **f = atom->f;
+
+    //initialize dynmat to all zeros
+    for (int i=0; i < dynlen; i++)
+        for (int j=0; j < dynlen; j++)
+            dynmat[i][j] = 0.;
+
+    energy_force(0);
+
+    if (comm->me == 0 && screen) fprintf(screen,"Calculating Dynamical Matrix...\n");
+
+    for (int i=1; i<=natoms; i++){
+        local_idx = atom->map(i);
+        //if (screen) fprintf(screen, "local idx = %i on proc %i with %i local\n",local_idx, comm->me, nlocal);
+        if (local_idx >= 0){
+            for (int alpha=0; alpha<3; alpha++){
+                displace_atom(local_idx, alpha, 1);
+                energy_force(0);
```

Comment:
`energy_force()` must be called by all MPI ranks, not only those owning the displaced atom.

Reply:
I have fallen for this trap enough times. I have fixed this.